engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/immutable/SliceArrayVectorNode.java

```diff
@@ -0,0 +1,60 @@
+package org.enso.interpreter.node.expression.builtin.immutable;
+
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.data.Array;
+import org.enso.interpreter.runtime.data.Vector;
+import org.enso.interpreter.runtime.error.PanicException;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+
+@BuiltinMethod(type = "Vector", name = "slice", description = "Returns a slice of this Vector.")
+public abstract class SliceArrayVectorNode extends Node {
+  SliceArrayVectorNode() {}
+
+  public static SliceArrayVectorNode build() {
+    return SliceArrayVectorNodeGen.create();
+  }
+
+  abstract Object execute(Object self, long start, long end);
+
+  @Specialization
+  Object executeArray(Array self, long start, long end) {
+    return Array.slice(self, start, end, self.length());
+  }
+
+  @Specialization
+  Object executeVector(
+      Vector self, long start, long end, @CachedLibrary(limit = "3") InteropLibrary iop) {
+    try {
+      return Array.slice(self, start, end, self.length(iop));
+    } catch (UnsupportedMessageException ex) {
+      CompilerDirectives.transferToInterpreter();
+      throw unsupportedMessageException(self);
+    }
+  }
+
+  @Specialization(replaces = {"executeArray", "executeVector"})
```

Comment:
The reason why I used own `Node` is: I wanted to use generic `self` - `Object` not an `Array`. I haven't found a way to do that with a DSL. Moreover there is
- #6299

we want to be able to have nodes that operate on any array-like object. It is not very useful in case of `slice` - we don't use it from `engine` itself. However we will benefit from having real nodes that specialize in case of `length` and `at`.